packages/webapp/src/containers/Map/useMaxZoom.js

```diff
@@ -1,22 +1,76 @@
 import { useDispatch, useSelector } from 'react-redux';
-import { mapCacheSelector, setMapCache } from './mapCacheSlice';
+import { mapCacheSelector, setMapCache, setRetrievedPoints } from './mapCacheSlice';
 import { userFarmSelector } from '../userFarmSlice';
 import { usePropRef } from '../../components/LocationPicker/SingleLocationPicker/usePropRef';
+import { pointSelector } from '../locationSlice';
+import { DEFAULT_MAX_ZOOM } from './constants';
 
 export function useMaxZoom() {
-  const { maxZoom } = useSelector(mapCacheSelector);
+  const { maxZoom, previousMaxZoom, retrievedPoints } = useSelector(mapCacheSelector);
   const { farm_id, grid_points } = useSelector(userFarmSelector);
   const dispatch = useDispatch();
-  const setMaxZoom = maxZoom => {
+  const setMaxZoom = (maxZoom) => {
     dispatch(setMapCache({ maxZoom, farm_id }));
   };
-  const getMaxZoom = maps => {
+  const points = useSelector(pointSelector);
+
+  const getMaxZoom = async (maps, map = null) => {
     if (!maxZoom) {
-      new maps.MaxZoomService()?.getMaxZoomAtLatLng(grid_points, (result) => {
-        if (result.status === 'OK') {
-          setMaxZoom(result.zoom);
+      const mapService = new maps.MaxZoomService();
+      const pointsToQuery = [];
+      const pointsCollections = [
+        { point: grid_points },
+        ...points.gate,
+        ...points.water_valve,
+        ...points.sensor,
+      ];
+      pointsCollections.forEach((element) => {
+        if (
+          !pointsToQuery.some(
+            (item) => item.lat === element.point.lat && item.lng === element.point.lng,
+          ) &&
+          !retrievedPoints.some(
+            (item) => item.lat === element.point.lat && item.lng === element.point.lng,
+          )
+        ) {
+          pointsToQuery.push(element.point);
         }
       });
+      const promises = pointsToQuery.map(function (point) {
+        return new Promise(function (resolve, reject) {
+          mapService?.getMaxZoomAtLatLng(point, function (result) {
```

Comment:
If there is no mapService, the case will be handled by the catch on the promises. I added a fix on that catch because it was setting the redux-store maxZoom but not the map maxZoom, now fixed, if there is a problem with mapService the DEFAULT_MAX_ZOOM is set inside the catch sentence.

Reply:
on error, if available the maxZoom will be retrieved from the stored cache of MaxZooms, if not, then DEFAULT_MAX_ZOOM will be used.